src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediatePreparedStatement.java

```diff
@@ -159,34 +176,283 @@ public IntermediatePortalStatement bind(
+        // that was used to determine the parameter types failed, so we can return a reasonable
+        // error
+        // message if the statement is invalid. If it is the unnamed statement or getting the param
+        // types succeeded, we will let the following EXECUTE message handle that, instead of
+        // sending
+        // the statement twice to the backend.
+        connection.analyzeUpdate(
+            Statement.of(this.parsedStatement.getSqlWithoutComments()), QueryAnalyzeMode.PLAN);
+      }
+      return new DescribeStatementMetadata(this.parameterDataTypes, null);
+    } finally {
+      if (columnsResultSet != null) {
+        columnsResultSet.close();
+      }
+    }
+  }
+
+  @VisibleForTesting
+  Statement transformToSelectParams(Set<String> parameters) {
+    switch (this.parsedStatement.getType()) {
+      case QUERY:
+        return transformSelectToSelectParams(parameters);
+      case UPDATE:
+        return transformDmlToSelectParams(parameters);
+      case CLIENT_SIDE:
+      case DDL:
+      case UNKNOWN:
+      default:
+        return Statement.of(this.parsedStatement.getSqlWithoutComments());
+    }
+  }
+
+  private Statement transformSelectToSelectParams(Set<String> parameters) {
+    return Statement.of(
+        String.format(
+            "select %s from (%s) p",
+            String.join(", ", parameters), this.parsedStatement.getSqlWithoutComments()));
+  }
+
+  private Statement transformDmlToSelectParams(Set<String> parameters) {
+    switch (getCommand(0)) {
+      case "INSERT":
+        return transformInsertToSelectParams(
+            this.connection, this.parsedStatement.getSqlWithoutComments(), parameters);
+      case "UPDATE":
+        return transformUpdateToSelectParams(
+            this.parsedStatement.getSqlWithoutComments(), parameters);
+      case "DELETE":
+        return transformDeleteToSelectParams(
+            this.parsedStatement.getSqlWithoutComments(), parameters);
+      default:
+        return null;
+    }
+  }
+
+  @VisibleForTesting
+  static @Nullable Statement transformInsertToSelectParams(
+      Connection connection, String sql, Set<String> parameters) {
+    SimpleParser parser = new SimpleParser(sql);
+    if (!parser.eat("insert")) {
+      return null;
+    }
+    parser.eat("into");
+    String table = parser.readIdentifier();
+    if (table == null) {
+      return null;
+    }
+    parser.skipWhitespaces();
+
+    List<String> columnsList = null;
+    if (parser.eat("(")) {
+      columnsList = parser.parseExpressionList();
+      if (!parser.eat(")")) {
+        return null;
+      }
+    }
+
+    parser.skipWhitespaces();
+    int potentialSelectStart = parser.getPos();
+    if (parser.eat("select")) {
+      // This is an `insert into <table> [(...)] select ...` statement. Then we can just use the
+      // select statement as the result.
+      return Statement.of(
```

Comment:
Can we make the Select statement method reusable?

Reply:
Done